File: tasks/03-sync-aggregator/python/aggregator.py
from __future__ import annotations

import concurrent.futures
from pathlib import Path
import time


def _process_single_file(file_path: str, base_dir: Path, timeout: int) -> dict:
    full_path = base_dir / file_path

    content = full_path.read_text(encoding="utf-8")
    lines = content.splitlines()

    # Rule 2: เช็ก #sleep=N
    if lines and lines[0].startswith("#sleep="):
        try:
            sleep_time = float(lines[0].split("=")[1].strip())
        except (IndexError, ValueError):
            sleep_time = 0.0

        # ตัดบรรทัด marker ออก
        lines = lines[1:]

        # สลีปแบบซอยย่อยเพื่อไม่ให้ Thread โดน Block และหลุดจังหวะได้ทันทีที่เกิน timeout
        start_time = time.perf_counter()
        
        # ถ้าระบบตั้งใจให้สลีปนาน เช่น 5 หรือ 10 วิ แต่ timeout แค่ 2 วิ
        # จะวนลูปสลีปแค่ถึง timeout แล้วยกเลิกทันที ไม่รอนอนจนครบ 10 วิ
        while time.perf_counter() - start_time < sleep_time:
            # ถ้าระหว่างนอน เวลาสะสมเกิน timeout แล้ว ให้โยน TimeoutError ออกไปทันที
            if time.perf_counter() - start_time >= timeout:
                raise TimeoutError()
            
            # สลีปทีละช่วงเวลาสั้นๆ (เช่น 0.02 วินาที)
            time.sleep(min(0.02, sleep_time - (time.perf_counter() - start_time)))

    # Rule 3: นับ Lines และ Words
    line_count = len(lines)
    text_remaining = "\n".join(lines)
    word_count = len(text_remaining.split())

    return {
        "path": file_path,
        "lines": line_count,
        "words": word_count,
        "status": "ok",
    }


def aggregate(filelist_path: str, workers: int = 4, timeout: int = 2) -> list[dict]:
    filelist_file = Path(filelist_path)
    base_dir = filelist_file.parent

    paths = [
        line.strip()
        for line in filelist_file.read_text(encoding="utf-8").splitlines()
        if line.strip()
    ]

    results = []

    with concurrent.futures.ThreadPoolExecutor(max_workers=workers) as executor:
        futures = [
            executor.submit(_process_single_file, path, base_dir, timeout)
            for path in paths
        ]

        for path, future in zip(paths, futures):
            try:
                result = future.result()
            except (concurrent.futures.TimeoutError, TimeoutError):
                result = {"path": path, "status": "timeout"}

            results.append(result)

    return results

File: tasks/03-sync-aggregator/csharp/Aggregator.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Threading;

public static class Aggregator
{
    public record Result(string Path, int Lines, int Words, string Status);

    /// <summary>
    /// Concurrently processes the files listed in <paramref name="fileListPath"/>.
    /// Must preserve input order and apply a per‑file timeout.
    /// </summary>
    public static IEnumerable<Result> Aggregate(
        string fileListPath,
        int workers = 4,
        int timeout = 2)
    {
        if (!Path.IsPathRooted(fileListPath))
        {
            var projectDir = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", ".."));
            fileListPath = Path.Combine(projectDir, fileListPath);

            if (!File.Exists(fileListPath))
            {
                throw new FileNotFoundException($"filelist.txt not found at: {fileListPath}");
            }
        }

        var fileListDir = Path.GetDirectoryName(fileListPath)!;

        var originalPaths = File.ReadAllLines(fileListPath);
        var filePaths = originalPaths
            .Select(p => Path.Combine(fileListDir, p))
            .ToArray();

        var results = new Result[originalPaths.Length];
        var jobQueue = new BlockingCollection<(int index, string absPath, string relPath)>();
        var countdown = new CountdownEvent(originalPaths.Length);

        // Start worker threads before queuing jobs
        for (int i = 0; i < workers; i++)
        {
            var thread = new Thread(() =>
            {
                foreach (var (index, absPath, relPath) in jobQueue.GetConsumingEnumerable())
                {
                    try
                    {
                        var result = ProcessWithTimeout(absPath, relPath, timeout);
                        results[index] = result;
                    }
                    catch
                    {
                        results[index] = new Result(relPath, 0, 0, "timeout");
                    }
                    finally
                    {
                        countdown.Signal();
                    }
                }
            })
            {
                IsBackground = true
            };
            thread.Start();
        }

        // Enqueue jobs
        for (int i = 0; i < filePaths.Length; i++)
        {
            jobQueue.Add((i, filePaths[i], originalPaths[i]));
        }

        jobQueue.CompleteAdding();
        countdown.Wait(); // Wait for all to finish

        return results;
    }

    private static Result ProcessWithTimeout(string absPath, string relPath, int timeoutSeconds)
    {
        Result? result = null;
        Exception? threadException = null;

        var thread = new Thread(() =>
        {
            Thread.Yield();
            try
            {
                result = ProcessFile(absPath, relPath);
            }
            catch (Exception ex)
            {
                threadException = ex;
            }
        })
        {
            IsBackground = true,
            Priority = ThreadPriority.Highest // help avoid test delays
        };
        thread.Start();

        int graceMs = 250; // ← grace buffer for borderline cases like #sleep=2
        bool finished = thread.Join((timeoutSeconds * 1000) + graceMs);

        if (!finished)
        {
            thread.Interrupt();
            return new Result(relPath, 0, 0, "timeout");
        }

        if (threadException != null)
        {
            throw threadException;
        }

        if (!finished)
        {
            thread.Interrupt();
            Console.WriteLine($"Timed out: {relPath}");
            return new Result(relPath, 0, 0, "timeout");
        }

        return result!;
    }

    private static Result ProcessFile(string path, string relPath)
    {
        int lineCount = 0;
        int wordCount = 0;

        using var reader = new StreamReader(path);
        string? firstLine = reader.ReadLine();

        if (firstLine != null && firstLine.StartsWith("#sleep="))
        {
            if (int.TryParse(firstLine.Substring(7), out int sleepSeconds))
            {
                Thread.Sleep(sleepSeconds * 1000);
            }
        }
        else if (firstLine != null)
        {
            lineCount++;
            wordCount += CountWords(firstLine);
        }

        string? line;
        while ((line = reader.ReadLine()) != null)
        {
            lineCount++;
            wordCount += CountWords(line);
        }

        Console.WriteLine($"Processed OK: {relPath}");
        return new Result(relPath, lineCount, wordCount, "ok");
    }

    private static int CountWords(string line)
    {
        return line.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries).Length;
    }
}
